cedar-lean/Cedar/Partial/Evaluator.lean

```diff
@@ -0,0 +1,159 @@
+-/
+
+import Cedar.Partial.Entities
+import Cedar.Partial.Expr
+import Cedar.Partial.Request
+import Cedar.Partial.Value
+import Cedar.Spec.Evaluator
+import Cedar.Spec.ExtFun
+import Cedar.Spec.Value
+
+/-! This file defines the semantics of Cedar partial evaluation. -/
+
+namespace Cedar.Partial
+
+open Cedar.Data
+open Cedar.Spec (Attr BinaryOp EntityUID Result UnaryOp intOrErr)
+open Cedar.Spec.Error
+
+def inₑ (uid₁ uid₂ : EntityUID) (es : Partial.Entities) : Bool :=
+  uid₁ == uid₂ || (es.ancestorsOrEmpty uid₁).contains uid₂
+
+def inₛ (uid : EntityUID) (vs : Set Spec.Value) (es : Partial.Entities) : Result Spec.Value := do
+  let uids ← vs.mapOrErr Spec.Value.asEntityUID .typeError
+  .ok (uids.any (Partial.inₑ uid · es))
+
+def apply₂ (op₂ : BinaryOp) (v₁ v₂ : Spec.Value) (es : Partial.Entities) : Result Partial.Value :=
+  match op₂, v₁, v₂ with
+  | .eq, _, _                                              => .ok (.value (v₁ == v₂))
+  | .less,   .prim (.int i), .prim (.int j)                => .ok (.value ((i < j): Bool))
+  | .lessEq, .prim (.int i), .prim (.int j)                => .ok (.value ((i ≤ j): Bool))
+  | .add,    .prim (.int i), .prim (.int j)                => intOrErr (i.add? j) >>= λ x => .ok (.value x)
+  | .sub,    .prim (.int i), .prim (.int j)                => intOrErr (i.sub? j) >>= λ x => .ok (.value x)
+  | .mul,    .prim (.int i), .prim (.int j)                => intOrErr (i.mul? j) >>= λ x => .ok (.value x)
+  | .contains,    .set vs₁, _                              => .ok (.value (vs₁.contains v₂))
+  | .containsAll, .set vs₁, .set vs₂                       => .ok (.value (vs₂.subset vs₁))
+  | .containsAny, .set vs₁, .set vs₂                       => .ok (.value (vs₁.intersects vs₂))
+  | .mem, .prim (.entityUID uid₁), .prim (.entityUID uid₂) => .ok (.value (Partial.inₑ uid₁ uid₂ es))
+  | .mem, .prim (.entityUID uid₁), .set (vs)               => Partial.inₛ uid₁ vs es >>= λ x => .ok (.value x)
+  | _, _, _                                                => .error .typeError
+
+def attrsOf (v : Spec.Value) (lookup : EntityUID → Result (Map Attr Partial.RestrictedValue)) : Result (Map Attr Partial.RestrictedValue) :=
+  match v with
+  | .record r              => .ok (r.mapOnValues Partial.RestrictedValue.value)
+  | .prim (.entityUID uid) => lookup uid
+  | _                      => .error typeError
+
+def hasAttr (v : Spec.Value) (a : Attr) (es : Partial.Entities) : Result Spec.Value := do
+  let r ← Partial.attrsOf v (fun uid => .ok (es.attrsOrEmpty uid))
+  .ok (r.contains a)
+
+def getAttr (v : Spec.Value) (a : Attr) (es : Partial.Entities) : Result Partial.RestrictedValue := do
+  let r ← Partial.attrsOf v es.attrs
+  r.findOrErr a attrDoesNotExist
+
+def bindAttr (a : Attr) (res : Result Partial.Value) : Result (Attr × Partial.Value) := do
+  let v ← res
+  .ok (a, v)
+
+def evaluate (x : Partial.Expr) (req : Partial.Request) (es : Partial.Entities) : Result Partial.Value :=
```

Comment:
In my experience, proofs are easier to structure if the code for each of these big match branches is in a separate function, whenever possible.  (This is why we broke up the concrete evaluator and typechecker code in this way.)

It becomes easier to write "inversion" lemmas that describe the behavior of the evaluation on a given case declaratively.  And these lemmas in turn make proofs easier to do.

YMMV of course :)

Reply:
Noted.  But let's take the current PR as-is and refactor in a future PR if it makes proofs easier?

Reply:
Was about to say this would be easier to read with functions for each case, but luckily I thought I'd check what this comment said first 😄

Reply:
Alright I that's several votes, so I will split it in a followup